src/qpx_mmc_css.cpp

```diff
@@ -0,0 +1,1909 @@
+ *
+ * \param dvdcss a \e libdvdcss instance.
+ * \param i_blocks an absolute block offset to seek to.
+ * \param i_flags #DVDCSS_NOFLAGS, optionally ored with one of #DVDCSS_SEEK_KEY
+ *        or #DVDCSS_SEEK_MPEG.
+ * \return the new position in blocks, or a negative value in case an error
+ *         happened.
+ *
+ * This function seeks to the requested position, in logical blocks.
+ *
+ * You typically set \p i_flags to #DVDCSS_NOFLAGS when seeking in a .IFO.
+ *
+ * If #DVDCSS_SEEK_MPEG is specified in \p i_flags and if \e libdvdcss finds it
+ * reasonable to do so (ie, if the dvdcss method is not "title"), the current
+ * title key will be checked and a new one will be calculated if necessary.
+ * This flag is typically used when reading data from a VOB.
+ *
+ * If #DVDCSS_SEEK_KEY is specified, the title key will be always checked,
+ * even with the "title" method. This is equivalent to using the now
+ * deprecated dvdcss_title() call. This flag is typically used when seeking
+ * in a new title.
+ */
+int seek_dvd(drive_info *drive, int lba, int flags) {
+  /* title cracking method is too slow to be used at each seek */
+  if (((flags & DVDCSS_SEEK_MPEG) &&
+       (drive->media.dvdcss.method != DVDCSS_METHOD_TITLE)) ||
+      (flags & DVDCSS_SEEK_KEY)) {
+    /* check the title key */
+    if (css_title(drive, lba)) {
+      return -1;
+    }
+  }
+
+  return seek(drive, lba);
+}
+
+/**
+ * \brief Read from the disc and decrypt data if requested.
+ *
+ * \param dvdcss a \e libdvdcss instance.
+ * \param p_buffer a buffer that will contain the data read from the disc.
+ * \param i_blocks the amount of blocks to read.
+ * \param i_flags #DVDCSS_NOFLAGS, optionally ored with #DVDCSS_READ_DECRYPT.
+ * \return the amount of blocks read, or a negative value in case an
+ *         error happened.
+ *
+ * This function reads \p i_blocks logical blocks from the DVD.
+ *
+ * You typically set \p i_flags to #DVDCSS_NOFLAGS when reading data from a
+ * .IFO file on the DVD.
+ *
+ * If #DVDCSS_READ_DECRYPT is specified in \p i_flags, dvdcss_read() will
+ * automatically decrypt scrambled sectors. This flag is typically used when
+ * reading data from a .VOB file on the DVD. It has no effect on unscrambled
+ * discs or unscrambled sectors, and can be safely used on those.
+ *
+ * \warning dvdcss_read() expects to be able to write \p i_blocks *
+ *          #DVDCSS_BLOCK_SIZE bytes in \p p_buffer.
+ */
+
+int read_dvd(drive_info *drive, unsigned char *data, int lba, int sector_count,
+             int flags)
+
+//	LIBDVDCSS_EXPORT int dvdcss_read ( dvdcss_t dvdcss, void *p_buffer,
+//                                          int i_blocks,
+//                                          int i_flags )
+{
+  unsigned char *p_buffer;
+  int i_ret = -1, i_index;
+
+  //	if (flags & DVDCSS_READ_DECRYPT) {
+  //		seek_dvd(drive,lba, DVDCSS_SEEK_KEY);
+  //	}
+
+  if (!read(drive, data, lba, sector_count)) {
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)